### PR TITLE
gh-122081: fixed crash in decimal.IEEEContext()

### PR DESCRIPTION
**Pull request by skirpichev**
Now

$ ./configure CFLAGS=-DEXTRA_FUNCTIONALITY -q && make -s && \
     ./python -m test test_decimal

- PASS

* Issue: gh-122081

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-07-21-10-45-24.gh-issue-122081.dNrYMq.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixed crash in optional :func:`!decimal.IEEEContext()` function.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-07-21-10-45-24.gh-issue-122081.dNrYMq.rst]: We should also document `decimal.IEEEContext` as well but in another PR.

Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-07-21-10-45-24.gh-issue-122081.dNrYMq.rst]:
```suggestion
Fix a crash in the :func:`!decimal.IEEEContext()` optional function
available via the ``EXTRA_FUNCTIONALITY`` configuration flag.
```

We should also document decimal.IEEEContext and `EXTRA_FUNCTIONALITY` as well but in another PR. Perhaps first document them and then move to your PR.

File: Modules/_decimal/_decimal.c [hunk 2]
```diff
@@ -1519,7 +1519,7 @@ init_extended_context(PyObject *v)
 #ifdef EXTRA_FUNCTIONALITY
 /* Factory function for creating IEEE interchange format contexts */
 static PyObject *
-ieee_context(PyObject *dummy UNUSED, PyObject *v)
+ieee_context(PyObject *self, PyObject *v)
```
Comment (picnixz) [inline Modules/_decimal/_decimal.c]:
```suggestion
ieee_context(PyObject *module, PyObject *v)
```

\+ change references in the function. `self` is a bit imprecise and could be thought as a "normal" object. I think it's fine to say that `self` here is the module.

Comment (skirpichev) [inline Modules/_decimal/_decimal.c]:
```suggestion
ieee_context(PyObject *Py_UNUSED(module), PyObject *v)
```
maybe this way?

Comment (picnixz) [reply Modules/_decimal/_decimal.c]: Don't you need the module to get it's state?

Comment (skirpichev) [reply Modules/_decimal/_decimal.c]: Oh, indeed)

But looking on the other module level function (e.g. PyDec_GetCurrentContext), I should reject your suggestion for consistency.

File: Misc/NEWS.d/next/Library/2024-07-21-10-45-24.gh-issue-122081.dNrYMq.rst (hunk 1 above)
Comment (skirpichev) [reply Misc/NEWS.d/next/Library/2024-07-21-10-45-24.gh-issue-122081.dNrYMq.rst]: here it's: https://github.com/python/cpython/pull/122003

**Review by picnixz — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2024-07-21-10-45-24.gh-issue-122081.dNrYMq.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Fix a crash in the :func:`!decimal.IEEEContext()` optional function
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2024-07-21-10-45-24.gh-issue-122081.dNrYMq.rst]:
```suggestion
Fix a crash in the :func:`!decimal.IEEEContext` optional function
```

**Review by kumaraditya303 — APPROVED** [review]

File: Modules/_decimal/_decimal.c (hunk 2 above)
Comment (kumaraditya303) [inline Modules/_decimal/_decimal.c]:
```suggestion
ieee_context(PyObject *module, PyObject *v)
```

File: Modules/_decimal/_decimal.c [hunk 4]
```diff
@@ -1536,7 +1536,7 @@ ieee_context(PyObject *dummy UNUSED, PyObject *v)
         goto error;
     }
 
-    decimal_state *state = get_module_state_by_def(Py_TYPE(v));
+    decimal_state *state = get_module_state(self);
```
Comment (kumaraditya303) [inline Modules/_decimal/_decimal.c]:
```suggestion
    decimal_state *state = get_module_state(module);
```

File: Modules/_decimal/_decimal.c (hunk 2 above)
Comment (picnixz) [reply Modules/_decimal/_decimal.c]: Actually, we wanted to keep the `self`. See https://github.com/python/cpython/pull/122082#discussion_r1720929120

Comment (skirpichev) [reply Modules/_decimal/_decimal.c]: Yeah, that was unfortunate change :(

Comment (skirpichev) [reply Modules/_decimal/_decimal.c]: @kumaraditya303, let me know if I should prepare a quick fix.

Comment (kumaraditya303) [reply Modules/_decimal/_decimal.c]: `ctxmanager_new` just uses `m` so it's not consistent, I would rather stick to module.

Comment (kumaraditya303) [reply Modules/_decimal/_decimal.c]: It's fine, I would rather stick to module, see my reasoning above.

**Comment by skirpichev** [comment]
@kumaraditya303, I'll double check, but I think that 3.12 backport - not needed.

**Comment by picnixz** [comment]
Removing backport to 3.12 per: https://github.com/python/cpython/issues/122081#issuecomment-2296119578

**Pull request merged**